### PR TITLE
Batch implementation with timer

### PR DESCRIPTION
**Pull request by FAYiEKcbD0XFqF2QK2E4viAHg8rMm2VbjYKdjTg**
Envoy supports Redis pipelined commands, but does not batch commands. By batching, I mean when non-pipelined or pipelined commands are sent through Envoy, Envoy should not send them upstream to Redis until a certain number of commands has been received, and then write them to the connection buffer for a particular upstream Redis in one go. See https://github.com/envoyproxy/envoy/issues/6365 for more details and discussion.

This PR implements a timer solution to the problem. The algorithm is as follows:

- If buffer is currently 0 and adding to buffer, set timer for x ms.
- If buffer is > 0, do nothing, timer is already set.
- If buffer is > X, immediate flush, cancel timer if set.

Batching is defaulted to off in the config- zero batch size, zero timer, so it should behave the same as the old way. Note that there is the extra overhead of a timer check in `ClientImpl::flushBufferAndResetTimer()` even if we have zero batch size- this may add slight overhead, but I have not been able to characterize the performance degradation in any meaningful way*. If there is degradation, or if desired by reviewers I can add a 3rd if statement to handle this in `client_impl.cc` ie `if (config_.maxBufferSizeBeforeFlush()==0) `  where we write without a timer check.

Besides updating unit tests, I've added 3 specific cases:

- Batching should be no-op with default settings
- Batching waits for timer fired when enabled (buffer size > request size)
- Buffer flushes when buffer size < sum(request sizes), and cancels flush timer

*I have done some evaluation on performance on my local VM- batching is definitely faster with single redis behind envoy. However, for real-world tests against our redis-benchmark machines in AWS we will have to wait until later this week.

### CODE REVIEW

**Comment by mattklein123** [comment]
@HenryYYang for first pass.

File: test/extensions/filters/network/common/redis/client_impl_test.cc [hunk 1]
```diff
@@ -99,6 +105,132 @@ class RedisClientImplTest : public testing::Test, public Common::Redis::DecoderF
...
+  // This means we do not batch requests, and thus the flush timer is never enabled.
+  InSequence s;
+
+  setup();
+
+  // Make the dummy request
+  Common::Redis::RespValue request1;
+  MockPoolCallbacks callbacks1;
+  EXPECT_CALL(*encoder_, encode(Ref(request1), _));
+  EXPECT_CALL(*flush_timer_, enabled());
```
Comment (HenryYYang) [inline test/extensions/filters/network/common/redis/client_impl_test.cc]: maybe make it explicit that enabled() returned false.

**Comment by FAYiEKcbD0XFqF2QK2E4viAHg8rMm2VbjYKdjTg** [comment]
@HenryYYang, yeah, saw version history in your PR when I was reviewing (docs/root/intro/version_history.rst), so I'll add a note there.

File: test/extensions/filters/network/common/redis/client_impl_test.cc [hunk 2]
```diff
@@ -99,6 +105,134 @@ class RedisClientImplTest : public testing::Test, public Common::Redis::DecoderF
...
+  Common::Redis::RespValue request1;
+  MockPoolCallbacks callbacks1;
+  EXPECT_CALL(*encoder_, encode(Ref(request1), _));
+  EXPECT_CALL(*flush_timer_, enableTimer(_));
+  PoolRequest* handle1 = client_->makeRequest(request1, callbacks1);
+  EXPECT_NE(nullptr, handle1);
+
+  // Pretend the the flush timer fires
+  EXPECT_CALL(*flush_timer_, enabled()).WillOnce(Return(false));
+  ; // timer is disabled, as it has already fired
```
Comment (FAYiEKcbD0XFqF2QK2E4viAHg8rMm2VbjYKdjTg) [inline test/extensions/filters/network/common/redis/client_impl_test.cc]: @mattklein123 this was the python formatter. Should I flag this as an issue?

**Comment by FAYiEKcbD0XFqF2QK2E4viAHg8rMm2VbjYKdjTg** [comment]
I suppose I'll get on the tip of mainline master again- that fixed coverage, but broke asan :|

File: api/envoy/config/filter/network/redis_proxy/v2/redis_proxy.proto [hunk 3]
```diff
@@ -46,6 +46,12 @@ message RedisProxy {
     // * '{user1000}.following' and '{user1000}.followers' **will** be sent to the same upstream
     // * '{user1000}.following' and '{user1001}.following' **might** be sent to the same upstream
     bool enable_hashtagging = 2;
+
+    // Maximum size of buffer before flush is triggered
```
Comment (mattklein123) [inline api/envoy/config/filter/network/redis_proxy/v2/redis_proxy.proto]: Can you specify what happens if this is unset? Same below?

File: docs/root/intro/version_history.rst [hunk 4]
```diff
@@ -59,6 +59,9 @@ Version history
 * redis: added :ref:`success and error stats <config_network_filters_redis_proxy_per_command_stats>` for commands.
 * redis: migrate hash function for host selection to `MurmurHash2 <https://sites.google.com/site/murmurhash>`_ from std::hash. MurmurHash2 is compatible with std::hash in GNU libstdc++ 3.4.20 or above. This is typically the case when compiled on Linux and not macOS.
 * redis: added :ref:`latency_in_micros <envoy_api_field_config.filter.network.redis_proxy.v2.RedisProxy.latency_in_micros>` to specify the redis commands stats time unit in microseconds.
+* redis: added
```
Comment (mattklein123) [inline docs/root/intro/version_history.rst]: Please merge master and move these docs to the 1.11.0 section.

**Comment by FAYiEKcbD0XFqF2QK2E4viAHg8rMm2VbjYKdjTg** [comment]
This is not ready for full review- I am attempting to do some performance testing in AWS first.

**Comment by FAYiEKcbD0XFqF2QK2E4viAHg8rMm2VbjYKdjTg** [comment]
Ok, had a chance to test in AWS. Results are less promising than hoped. Perhaps my understanding of the `rpc-perf` script is imperfect, but I'm only seeing around an 8% perf improvement (the rate metric below), but a big improvement in p99 latency.

TODO: Have to also add in notes on when unset per earlier comment from Matt.

# Test

## Procedure

Using an appropriate config file for Envoy (depending on if we are batching or not, ie we add batch params):

	$ bazel-bin/source/exe/envoy-static -c [CONFIG] --service-cluster redisbenchmark-staging-iad --service-node nico-vm -l critical
	$ redis-cli -h 172.31.43.35 FLUSHALL (or )
	$ ./rpc-perf -s 127.0.0.1:6380 --config basic.toml -p redis | python3 process_rpc_perf_output.py

## Basic Configuration of Toml file

	request-timeout = 200
	connect-timeout = 500
	threads = [Threads]
	connections = [Connections]
	duration = 20
	windows = 5
	protocol = "redis"
	tcp-nodelay = true
	ipv4 = true
	ipv6 = true
	database = 0

	[[workload]]
	name = "get"
	method = "get"
	rate = [Request Rate]
	  [[workload.parameter]]
	  style = "random"
	  size = 64
	  regenerate = true

## EC2 -> EC2

First, we compare performance of batching for different request rates to another EC2 instance running Redis.

### Test 1

	threads 5
	connections 5
	request rate 100000
	flush_buffer length 1024
	flush_buffer timeout 3ms

#### Non-batch

rates: 15212.6 rps
success: 100 %
p50: 5625.6 ms
p99: 28655.2 ms

#### Batch

rates: 17425 rps
success: 100 %
p50: 5599.6 ms
p99: 10559 ms


###  Test 2

	threads 5
	connections 5
	request rate 300000
	flush_buffer length 1024
	flush_buffer timeout 3ms

#### Non-batch

rates: 13620 rps
success: 100 %
p50: 6936.8 ms
p99: 21326.6 ms

#### Batch

rates: 14822 rps
success: 100 %
p50: 5355.6 ms
p99: 21689 ms

## EC2 -> Elasticache

### Test 1

	threads 5
	connections 5
	request rate 100000
	flush_buffer length 1024
	flush_buffer timeout 3ms

#### Non-batch

rates: 15294.2 rps
success: 100 %
p50: 5479.8 ms
p99: 27688.6 ms

#### Batch

rates: 16511.2 rps
success: 100 %
p50: 6444.6 ms
p99: 8201.2 ms

### Test 2

	threads 20
	connections 20
	request rate 100000
	flush_buffer length 1024
	flush_buffer timeout 3ms
	
#### Non-batch

rates: 14373.8 rps
success: 100 %
p50: 23098.4 ms
p99: 78087.4 ms

#### Batch

rates: 14660.8 rps
success: 100 %
p50: 23511.2 ms
p99: 74195 ms

**Comment by mattklein123** [comment]
@FAYiEKcbD0XFqF2QK2E4viAHg8rMm2VbjYKdjTg so you want to review this? Or you are still working on it? If so, PTAL at CI.

/wait

**Comment by FAYiEKcbD0XFqF2QK2E4viAHg8rMm2VbjYKdjTg** [comment]
@mattklein123 sure, let's see if my changes fix the CI issues. Mitch's redirection work came through on merge and broke builds of some tests. However, if time is tight I'd prefer a review on https://github.com/envoyproxy/envoy/pull/6446, which has been updated per your comments.

Originally I had planned on waiting for Henry and doing more testing, but I think the above results are pretty straightforward, and little perf bump is not unwelcome.

File: api/envoy/config/filter/network/redis_proxy/v2/redis_proxy.proto [hunk 5]
```diff
@@ -53,6 +53,13 @@ message RedisProxy {
     // need to be known to the cluster manager. If the command cannot be redirected, then the
     // original error is passed downstream unchanged. By default, this support is not enabled.
     bool enable_redirection = 3;
+
+    // Maximum size of buffer before flush is triggered. If this is unset, the buffer flushes
+    // whenever it receives data and performs no batching.
+    uint32 max_buffer_size_before_flush = 4;
+
+    // Buffer is flushed every N milliseconds (unless the buffer is filled before the timer fires)
```
Comment (mattklein123) [inline api/envoy/config/filter/network/redis_proxy/v2/redis_proxy.proto]: Can you expand? What buffer? How does it relate to `max_buffer_size_before_flush`? Also what is the behavior when this is unset?

File: api/envoy/config/filter/network/redis_proxy/v2/redis_proxy.proto [hunk 6]
```diff
@@ -53,6 +53,13 @@ message RedisProxy {
     // need to be known to the cluster manager. If the command cannot be redirected, then the
     // original error is passed downstream unchanged. By default, this support is not enabled.
     bool enable_redirection = 3;
+
+    // Maximum size of buffer before flush is triggered. If this is unset, the buffer flushes
```
Comment (mattklein123) [inline api/envoy/config/filter/network/redis_proxy/v2/redis_proxy.proto]: Can you add a little color on when someone might want to set this vs. not set it, and what a good value would be if they want to try it?

File: source/extensions/filters/network/common/redis/client.h [hunk 7]
```diff
@@ -110,6 +110,16 @@ class Config {
    * processed.
    */
   virtual bool enableRedirection() const PURE;
+
+  /**
+   * @return buffer size for batching commands for a single upstream host.
+   */
+  virtual unsigned int maxBufferSizeBeforeFlush() const PURE;
```
Comment (mattklein123) [inline source/extensions/filters/network/common/redis/client.h]: s/int/uint32_t

File: source/extensions/filters/network/common/redis/client_impl.cc [hunk 8]
```diff
@@ -11,7 +11,10 @@ ConfigImpl::ConfigImpl(
     const envoy::config::filter::network::redis_proxy::v2::RedisProxy::ConnPoolSettings& config)
     : op_timeout_(PROTOBUF_GET_MS_REQUIRED(config, op_timeout)),
       enable_hashtagging_(config.enable_hashtagging()),
-      enable_redirection_(config.enable_redirection()) {}
+      enable_redirection_(config.enable_redirection()),
+      max_buffer_size_before_flush_(
+          config.max_buffer_size_before_flush()), // This is a scalar, so default is zero.
+      buffer_flush_timeout_(PROTOBUF_GET_MS_OR_DEFAULT(config, buffer_flush_timeout, 3)) {}
```
Comment (mattklein123) [inline source/extensions/filters/network/common/redis/client_impl.cc]: Please document 3ms default, and presumably this doesn't take effect until max_buffer_size_before_flush_ > 0?

File: source/extensions/filters/network/common/redis/client_impl.cc [hunk 9]
```diff
@@ -48,12 +52,26 @@ ClientImpl::~ClientImpl() {
...
+  if (flush_timer_->enabled()) {
+    flush_timer_->disableTimer();
+  }
+  connection_->write(encoder_buffer_, false);
+}
+
 PoolRequest* ClientImpl::makeRequest(const RespValue& request, PoolCallbacks& callbacks) {
   ASSERT(connection_->state() == Network::Connection::State::Open);
 
+  bool empty_buffer = encoder_buffer_.length() == 0;
```
Comment (mattklein123) [inline source/extensions/filters/network/common/redis/client_impl.cc]: nit: const

File: source/extensions/filters/network/common/redis/client_impl.cc [hunk 10]
```diff
@@ -48,12 +52,26 @@ ClientImpl::~ClientImpl() {
...
+
   pending_requests_.emplace_back(*this, callbacks);
   encoder_->encode(request, encoder_buffer_);
-  connection_->write(encoder_buffer_, false);
+
+  if (encoder_buffer_.length() > config_.maxBufferSizeBeforeFlush()) {
+    flushBufferAndResetTimer();
+  } else if (empty_buffer) {
+    flush_timer_->enableTimer(std::chrono::milliseconds(config_.bufferFlushTimeoutInMs()));
+  } // else keep adding to buffer
```
Comment (mattklein123) [inline source/extensions/filters/network/common/redis/client_impl.cc]: ?

File: source/extensions/filters/network/common/redis/client_impl.h [hunk 11]
```diff
@@ -41,11 +41,17 @@ class ConfigImpl : public Config {
   std::chrono::milliseconds opTimeout() const override { return op_timeout_; }
   bool enableHashtagging() const override { return enable_hashtagging_; }
   bool enableRedirection() const override { return enable_redirection_; }
+  unsigned int maxBufferSizeBeforeFlush() const override { return max_buffer_size_before_flush_; }
```
Comment (mattklein123) [inline source/extensions/filters/network/common/redis/client_impl.h]: uint32_t, same elsewhere

File: test/extensions/filters/network/common/redis/client_impl_test.cc [hunk 12]
```diff
@@ -64,6 +64,11 @@ class RedisClientImplTest : public testing::Test, public Common::Redis::DecoderF
     upstream_connection_ = new NiceMock<Network::MockClientConnection>();
     Upstream::MockHost::MockCreateConnectionData conn_info;
     conn_info.connection_ = upstream_connection_;
+
+    // Create timers in order they are created in client_impl.cc
+    connect_or_op_timer_ = {new Event::MockTimer(&dispatcher_)};
```
Comment (mattklein123) [inline test/extensions/filters/network/common/redis/client_impl_test.cc]: nit: now braces needed here and next line.

File: test/extensions/filters/network/common/redis/client_impl_test.cc (hunk 2 above)
Comment (mattklein123) [reply test/extensions/filters/network/common/redis/client_impl_test.cc]: What do you mean?

File: test/extensions/filters/network/common/redis/client_impl_test.cc [hunk 13]
```diff
@@ -107,6 +242,8 @@ TEST_F(RedisClientImplTest, Basic) {
   Common::Redis::RespValue request1;
   MockPoolCallbacks callbacks1;
   EXPECT_CALL(*encoder_, encode(Ref(request1), _));
+  EXPECT_CALL(*flush_timer_, enabled()).WillOnce(Return(false));
+  ;
```
Comment (mattklein123) [inline test/extensions/filters/network/common/redis/client_impl_test.cc]: del, same elsewhere.

Comment (FAYiEKcbD0XFqF2QK2E4viAHg8rMm2VbjYKdjTg) [reply test/extensions/filters/network/common/redis/client_impl_test.cc]: Ah yes, the format checker added those. I'll remove them and ping OSS envoy slack if happens again.

**Comment by FAYiEKcbD0XFqF2QK2E4viAHg8rMm2VbjYKdjTg** [comment]
re `// timer is disabled, as it has already fired` and what is means, I cannot add a reply to that thread for whatever reason. What I meant for that line is this: Inside the timer callback function, we check if the timer is enabled and if so, cancel it- this is for the case where we fill the buffer before the timer fires. Thus, we expect the timer.enabled() call to return false inside this callback if the timer has already fired. I'll expand the comment to explain this.

File: test/extensions/filters/network/common/redis/client_impl_test.cc (hunk 12 above)
Comment (FAYiEKcbD0XFqF2QK2E4viAHg8rMm2VbjYKdjTg) [reply test/extensions/filters/network/common/redis/client_impl_test.cc]: ok

File: source/extensions/filters/network/common/redis/client_impl.cc (hunk 10 above)
Comment (FAYiEKcbD0XFqF2QK2E4viAHg8rMm2VbjYKdjTg) [reply source/extensions/filters/network/common/redis/client_impl.cc]: In a previous line `encoder_->encode(request, encoder_buffer_);`, we buffer data. This comment means that the third option is to keep buffering, though without a clause. I'll reword the comment.

File: source/extensions/filters/network/common/redis/client_impl.cc (hunk 8 above)
Comment (FAYiEKcbD0XFqF2QK2E4viAHg8rMm2VbjYKdjTg) [reply source/extensions/filters/network/common/redis/client_impl.cc]: Correct, if the buffer is always flushed the timer is never used.

File: api/envoy/config/filter/network/redis_proxy/v2/redis_proxy.proto (hunk 5 above)
Comment (FAYiEKcbD0XFqF2QK2E4viAHg8rMm2VbjYKdjTg) [reply api/envoy/config/filter/network/redis_proxy/v2/redis_proxy.proto]: Added comments

**Comment by FAYiEKcbD0XFqF2QK2E4viAHg8rMm2VbjYKdjTg** [comment]
@mattklein123 ready for review (your comments addressed, integration test added). 

Note that asan/tsan have failed on unrelated tests- for one build `//test/integration:hds_integration_test`, on another build `//test/integration:http_timeout_integration_test`- so maybe there is some integration test flakiness going on? Perhaps related to Google outage?

File: api/envoy/config/filter/network/redis_proxy/v2/redis_proxy.proto [hunk 14]
```diff
@@ -59,6 +59,27 @@ message RedisProxy {
...
+
+    // The encoded request buffer is flushed N milliseconds after the first request has been
+    // encoded, unless the buffer size has already exceeded `max_buffer_size_before_flush`.
+    // If `max_buffer_size_before_flush` is not set, this flush timer is not used. Otherwise,
+    // the timer should be set according to the number of clients, overall request rate and
+    // desired maximum latency for a single command. For example, if there are many requests
+    // being batched together at a high rate, the buffer will likely be filled before the timer
+    // fires. Alternatively, if the request rate is lower the buffer will not be filled as often
+    // before the timer fires.
+    // If `max_buffer_size_before_flush` is set, but `buffer_flush_timeout`, the latter
```
Comment (mattklein123) [inline api/envoy/config/filter/network/redis_proxy/v2/redis_proxy.proto]: "but buffer_flush_timeout is not," ?

File: source/extensions/filters/network/common/redis/client_impl.cc [hunk 15]
```diff
@@ -48,12 +56,27 @@ ClientImpl::~ClientImpl() {
...
   ASSERT(connection_->state() == Network::Connection::State::Open);
 
+  const bool empty_buffer = encoder_buffer_.length() == 0;
+
   pending_requests_.emplace_back(*this, callbacks);
   encoder_->encode(request, encoder_buffer_);
-  connection_->write(encoder_buffer_, false);
+
+  // If buffer is full, flush. If the the buffer is empty, start the timer.
+  if (encoder_buffer_.length() > config_.maxBufferSizeBeforeFlush()) {
```
Comment (mattklein123) [inline source/extensions/filters/network/common/redis/client_impl.cc]: technically I think this should be >= ?

File: source/extensions/filters/network/common/redis/client_impl.cc [hunk 16]
```diff
@@ -48,12 +56,27 @@ ClientImpl::~ClientImpl() {
...
 PoolRequest* ClientImpl::makeRequest(const RespValue& request, PoolCallbacks& callbacks) {
   ASSERT(connection_->state() == Network::Connection::State::Open);
 
+  const bool empty_buffer = encoder_buffer_.length() == 0;
+
   pending_requests_.emplace_back(*this, callbacks);
   encoder_->encode(request, encoder_buffer_);
-  connection_->write(encoder_buffer_, false);
+
+  // If buffer is full, flush. If the the buffer is empty, start the timer.
```
Comment (mattklein123) [inline source/extensions/filters/network/common/redis/client_impl.cc]: nit: "was empty before this request"

Comment (FAYiEKcbD0XFqF2QK2E4viAHg8rMm2VbjYKdjTg) [reply source/extensions/filters/network/common/redis/client_impl.cc]: yeah, that's a much better way of stating it.

**Comment by FAYiEKcbD0XFqF2QK2E4viAHg8rMm2VbjYKdjTg** [comment]
@mattklein123 fixed things according to your comments; all checks now pass on latest master.

**Pull request merged**